Gesture typing ai

EtienneDesticourt opened the pull request:
The current framework for gesture typing is a good starting point for the feature but isn't accurate enough to be a valid alternative to classical finger typing.

# Current Algorithm

The current system uses a corner matching algorithm to classify gestures. The Keyboard object starts by loading a dictionary and feeds it to the gesture detector object which then generates paths of corners for every word. To do so it first creates a sequence of points by getting the location of the center of each key needed to type the word, then it keeps those points that are at an angle > 170° from each others.

Afterward, when the user traces a gesture, the points the gesture goes through are passed to the gesture detector through keyboard events. Once the gesture is over, only the corners (angles > 170°) are kept. Then the dictionary if filtered to keep only the words who start by the key that was pressed at the start of the user gesture. A cumulative euclidean distance is then calculated between the corners of those words and the corners of the traced gesture.

We keep the n-lowest distances for suggestions, with the lowest being the typed word.

# Identifying the limitations of the current algorithm

## Loss of information

The corner matching algorithm ignores all the data in the gesture except the corner, this means getting rid of a lot of useful information. The most flagrant example is when it's applied to words where there are two successive identical letters.

![image](https://user-images.githubusercontent.com/8978248/67991760-a3ec8400-fc3a-11e9-9998-fc9d165b5e87.png)

## Untypable words

Some words have exactly the same corners which mean only the most frequent one out of this set of words will be typable using this algorithm. The others will never be selected correctly.

![image](https://user-images.githubusercontent.com/8978248/67991905-2a08ca80-fc3b-11e9-8faa-77ea02264d6e.png)
![image](https://user-images.githubusercontent.com/8978248/67991972-53295b00-fc3b-11e9-8334-9070f2b8ea26.png)

## Inability to match imperfect patterns

The corner matching algorithm often has trouble matching smooth gestures as well as those that aren't perfectly centered on the keys.

![image](https://user-images.githubusercontent.com/8978248/67992769-4a865400-fc3e-11e9-84ec-83e97075ebe9.png)


# Data gathering

I've added a log trace to the gesture keyboard at https://github.com/EtienneDesticourt/AnySoftKeyboard/commit/470483fc652ada8027847aa264f5964cf80e606d#diff-3e4aa7a298643ae5be270bde873a45a7R450 so we can save the gestures as they're typed. 
I've started to build a dataset but it will be biased toward my typing patterns if I'm the only one contributing to it. Ideally we should get a few people to record their gestures and share the logs.

It would be good to setup a data recording protocol as well: agree on a set of sentences that contain enough of the most frequent words, as well as all sort of gesturing patterns (left to right, right to left, top to bottom, etc...). It's likely that the previous word as well as the following word influence the gesture for a given word so recording single words would not be as representative of real typing as full sentences.

Ideally we would need multiple recordings of the same sentence for a single individual to study the standard deviation in a single pattern.

Below is an example of me typing the word "keyboard" as carefully as I can, it's very rough despite the great care I took:
![image](https://user-images.githubusercontent.com/8978248/67992588-a43a4e80-fc3d-11e9-9c5a-1caea29cbad1.png)


# Pattern Characterization

The noise patterns of real gestures need to be analyzed.  A short EDA on the dataset should be enough to give us all the information we need to proceed with model building.

**_I haven't started this step yet._**

# Parametric model selection and implementation

A parametric model has to be selected. Deep learning is overkill for now, there isn't enough data and it's more battery intensive, there are some low hanging fruits to pick beforehand.

_**I haven't started this step yet.**_

menny commented:
@EtienneDesticourt you can fix the formatting issues by running `./gradlew :googleJavaFormat`

menny commented on app/src/main/java/com/anysoftkeyboard/ime/AnySoftKeyboardWithGestureTyping.java:
```diff
@@ -383,6 +447,7 @@ public void onGestureTypingInputDone() {
 
         final GestureTypingDetector currentGestureDetector = mCurrentGestureDetector;
         if (ic != null && currentGestureDetector != null) {
+            Logger.d(TAG, "Completed gesture: %s.", currentGestureDetector.getWorkspaceToString());
```
For privacy (and ease-of-use) concerns, I suggest that:
* should be activated by a shared-prefs setting (disabled default).
* should be disabled if in _Incognito_ mode.
* this will be written to a file in the public external storage.

If users will be able to enable this with a click and then easily find the log file that would make it more likely they will contribute.

EtienneDesticourt replied:
Good point. I meant for it to be used by developers while this pull request is open (with abd logcat), and to remove it in the last commit before merging. I don't know that I'll have time to implement UI changes as I'm not familiar with the codebase at all and Android is not my strong suit anymore so I'll focus on modeling for now and get back to this right before merging.

menny commented:
This documentation is priceless.

ildar commented:
Anybody have the apk? Actions already purged the apk-s that were built.
Or how to kick github to rebuild it? I'd want to test.

menny commented:
@ildar I don't think that this branch is ready for testers yet.

menny commented:
still working on this.

Jackalaq commented:
Great to hear. Thanks for all of your work; I'll try and help, although i live offgrid eighth limited 12v power and can't program any more. I not be that much use...

s-c-p commented:
@EtienneDesticourt 
Wow, wow, wow :1st_place_medal:  :+1: 

(sorry, had to comment, I am struggling to finish ML course, and all THIS in JAVA! You must be in top 10 smartest person on earth right now)

EtienneDesticourt commented:
Thanks @s-c-p, sadly I am dealing with severe health issues and haven't been able to dedicate any time to it these past few months. It's already functional, I use it daily on my phone but the normalization steps must be added so that it works with displays of any size. Basically the keyboard coordinates must be normalized so that they lie between 0 and 1 and the gestures must use this coordinate system, and the algorithm's parameters must be adjusted to work in that system. Once that's done it should be fully functional.

ildar commented:
I managed to build this and I'm sorry to confirm @menny 's words that it's
too early for testers: it just hangs for some time (up to a few seconds)
after every gesture.
@EtienneDesticourt did a great job, thanks a lot. But now we just hope it
won't vanish.

jeremiah-miller commented:
Sorry if this is an ignorant question, I don't really understand how ASK's gesture typing works or how the flow keyboards gesture typing works, but I do know that the gesture typing in the flow keyboard is really smooth and works great, maybe ASK could use a similar algorithm? 
https://sourceforge.net/p/flowkeyboard/code/HEAD/tree/ 
#2545

ildar commented:
@jeremiah-miller thank you for referencing.
ASK's definitely stuck with gesture typing, any related info is appropriate.

menny commented:
Could you add a markdown file with information about the algorithm you picked (Shark2)? And anything you'd like to mention about the actual implementation? This `REAME.md`  file should be located at `ime/app/src/main/java/com/anysoftkeyboard/gesturetyping/README.md`
It shouldn't be too long, just high level, and maybe linking to further reading. And a link to this PR.

Also, add a small section in the README.md file at the root about Gesture-Typing?

EtienneDesticourt commented:
I've added a readme and rebased again to fix some issues (sorry hope you weren't working on it already). To be clear it's not based on Shark2, I looked into some articles on it when I started working on our implementation but our architecture ended up completely different. Our pruning is completely custom, the way we generate our gestures is different and the algorithm itself is quite different apart from some obvious similarities (using the euclidean distance and normalizing the gestures which are both common sense measures). It's all in the readme.

I wasn't sure what you wanted to add to the root README so I haven't added anything.

khimaros commented:
are there any remaining changes required before this can be merged?

Wintandre commented:
I've been testing this out on an old Galaxy with Android 8 and it works great, at least in terms of the suggestions. It even detects words like 'even' and 'never', which the published release currently struggles with. The only issues are speed of matching the suggestions which is sometimes slow and I experienced a single crash.

Really great work from @EtienneDesticourt

TRSx80 commented:
@EtienneDesticourt,

Brilliant work, mate!  This is the handiest thing since sliced bread!  :smile: :beers: 

@ all,

I have been playing with my new PinePhone[0] recently, and one of biggest pain points for me is the lack of gesture typing.  I am (very) happy to be off Android and on a real Linux device, however gesture typing is one thing I really miss.

So, perhaps these are naive questions, but...

How much work would be involved in...

1.0. ...splitting this off into its own separate library (so that multiple F/LOSS keyboard projects might leverage it)?

2.1. ...adapting this to work with any of keyboards which are currently already available on Linux?

(and/or)

2.2. ...adapting AnySoftKeyboard itself to regular GNU/Linux (not Android).

Perhaps big questions, I know, and forgive me if they are ignorant.  But based on my research thus far, it seems like this is one of (if not the most) viable F/LOSS gesture typing implementation I have come across yet.

Any information would be appreciated.

I don't want to put the cart before the horse, but I would be very interested in financially sponsoring such work, should it turn out to be feasible and someone have the time and interest to work on it.

Cheers,
TRS-80

[0] For those who may be unaware, it's an actual, no bee ess GNU/Linux (not Android!) smartphone ([wiki](https://wiki.pine64.org/wiki/PinePhone), [store](https://pine64.com/product/pinephone-beta-edition-with-convergence-package-linux-smartphone-open-for-preorder-soon/))!

ugirishm commented:
This has some performance issue. Try gliding for "swipe", the gesture stops on "e" for few seconds before the suggestions are shown and many times it suggests "swore" instead of "swipe".

bugnano commented:
I have 2 questions regarding this:

1. The last comment in this PR dates back to 2021. Is this still being worked on?
2. Is it only working for the English language, or it can work on other languages as well?

Thank you.

SuperPauly commented:
@menny can it be pushed? Super useful feature.

bull500 commented:
@EtienneDesticourt i am in awe! The write up and the thought process is *chef's kiss. :100: :exploding_head: 
Thank you for bring this to the open source world. @EtienneDesticourt @menny 

I saw your un-typable(feel, fell) words section and i felt that i should share this:
Early on there was a keyboard called swype and i think they were they first to introduce swipe typing/glide typing to android
Their approach was to circle around the letter and move ahead. so f-circle on e-l -> feel

It might be a older approach but i think it could help
Hoping it aids/helps you in bring this tech to the FLOSS world

I've linked an old YouTube video below which shows it in action (:20 seconds - exact timestamp)
https://www.youtube.com/watch?v=pLOrr_-twwA

matheuswillder commented:
> I haven't use AnySwift since that time myself but iirc it did at some point add Gesture typing tho so you should be good to go.

When this PR was created, gesture typing was already implemented in AnySoftKeyboard, but considered beta/experimental and therefore disabled by default, a status it maintains *to this day*.

This PR was looking to replace the existing implementation with a better one, so it's a shame it was never merged, gesture typing is a feature that open source keyboards leaves something to be desired *to this day*.

I hope @EtienneDesticourt is well and aware that the effort and dedication put in this PR is appreciated, even if it was never merged.

matheuswillder commented:
@SavSanta:
> I haven't use AnySwift since that time myself but iirc it did at some point add Gesture typing tho so you should be good to go.

When this PR was created, gesture typing was already implemented in AnySoftKeyboard, but considered beta/experimental and therefore disabled by default, and it continues like this *to this day*.

This PR was looking to replace the existing implementation with a better one, so it's a shame it was never merged, because gesture typing is a feature that open source keyboards still leaves something to be desired *to this day*.

@EtienneDesticourt:
>Thanks @s-c-p, sadly I am dealing with severe health issues and haven't been able to dedicate any time to it these past few months.

I hope you are well and that you are aware that your effort and dedication put in this PR are appreciated, even if it was never merged.
